### DISCUSSION TITLE
how to export updates in runfiles from a target ?

### DISCUSSION THREAD

**Discussion by taozhou-glean** [bazelbuild/bazel#18486] [Q&A]
we have a bazel setup as follows:

```
load("@npm//@lingui/cli:index.bzl", lingui = "lingui")
load("@npm//webpack-cli:index.bzl", webpack = "webpack_cli")


lingui(
    name = "lingui_web_srcs_with_pseudo",
    args = [],
    data = [
        "//web:srcs",
    ] + some_other_deps,
)

webpack(
  name = "webpack",
  data = [
        ":lingui_web_srcs_with_pseudo",
        ....
  ]
)
```

the idea is that before piping the `web:srcs` to webpack, we want to run `lingui` (which is an i18n framework) to generate the message file for pseudo locale (we don't want to check pseudo locale to source..)

but as `lingui` here only updates the runfiles within its run, so the change it made never propagates to other targets, and the rule itself does not produce any outputs either....

I have tried to use a `genrule` to copy and export files from `lingui`, but could not find where I can retrieve the path to the runfiles...

(here we are using the rules_js, but I'm posting here as rules_js seems to be deprecated and also this seems to be a general question regarding how to include generated artifacts that exists in runfiles... or how to avoid having it in runfiles at the beginning...)

**Comment by sluongng** [comment]
> here we are using the rules_js, but I'm posting here as rules_js seems to be deprecated 

Are you referring to https://github.com/bazelbuild/rules_nodejs or https://github.com/aspect-build/rules_js?

It would be helpful if you could provide the specific version of the rules you use.

> the idea is that before piping the web:srcs to webpack, we want to run lingui (which is an i18n framework) to generate the message file for pseudo locale (we don't want to check pseudo locale to source..)
> 
> but as lingui here only updates the runfiles within its run, so the change it made never propagates to other targets, and the rule itself does not produce any outputs either....

So I think both `lingui` and `webpack` here are simple wrappers around `nodejs_binary` of the dependencies?
They perform pretty similarly to `genrule` where they operate on top of some data, then generate outputs according to `outs` attribute.
Do you have the `outs` attribute set on these rules?

In general, I would advise you to try and read the `index.bzl` files generated and see what it does. This command should give you the location of the file to open it and read through it.
```
bazel query --output=location @npm//@lingui/cli:index.bzl
```
